File: LiveData/app/src/main/java/com/kurio/mvvmtesting/db/NoteRepo.java
package com.kurio.mvvmtesting.db;

import android.app.Application;
import android.arch.lifecycle.LiveData;
import android.os.AsyncTask;

import com.kurio.mvvmtesting.model.Note;

import java.util.List;

public class NoteRepo {
    private NoteDao noteDao;
    private LiveData<List<Note>> allNotes;

    public NoteRepo(Application application) {
        noteDao = NoteRoomDatabase.getInstance(application).noteDao();
        allNotes = noteDao.getAllNote();
    }

    public void insertNote(Note note) {
        new InsertAsyncTask(noteDao).execute(note);
    }

    public LiveData<List<Note>> getAllNotes() {
        return allNotes;
    }

    public void updateNote(Note note) {
        new UpdateAsyncTask(noteDao).execute(note);
    }

    public void deleteNote(Note note) {
        new DeleteAsyncTask(noteDao).execute(note);
    }

    private static class InsertAsyncTask extends AsyncTask<Note, Void, Void> {
        private NoteDao noteDao;

        InsertAsyncTask(NoteDao noteDao) {
            this.noteDao = noteDao;
        }

        @Override
        protected void onPreExecute() {
            super.onPreExecute();
        }

        @Override
        protected Void doInBackground(Note... notes) {
            noteDao.insert(notes[0]);
            return null;
        }

        @Override
        protected void onPostExecute(Void aVoid) {
            super.onPostExecute(aVoid);
        }
    }

    private static class UpdateAsyncTask extends AsyncTask<Note, Void, Void> {
        private NoteDao noteDao;

        UpdateAsyncTask(NoteDao noteDao) {
            this.noteDao = noteDao;
        }

        @Override
        protected void onPreExecute() {
            super.onPreExecute();
        }

        @Override
        protected Void doInBackground(Note... notes) {
            noteDao.update(notes[0]);
            return null;
        }

        @Override
        protected void onPostExecute(Void aVoid) {
            super.onPostExecute(aVoid);
        }
    }

    private static class DeleteAsyncTask extends AsyncTask<Note, Void, Void> {
        private NoteDao noteDao;

        DeleteAsyncTask(NoteDao noteDao) {
            this.noteDao = noteDao;
        }

        @Override
        protected void onPreExecute() {
            super.onPreExecute();
        }

        @Override
        protected Void doInBackground(Note... notes) {
            noteDao.delete(notes[0]);
            return null;
        }

        @Override
        protected void onPostExecute(Void aVoid) {
            super.onPostExecute(aVoid);
        }
    }

}

File: LiveData (Kotlin)/app/src/main/java/com/kurio/mvvmtesting/db/NoteRepo.kt
package com.kurio.mvvmtesting.db

import android.app.Application
import android.arch.lifecycle.LiveData
import android.os.AsyncTask

import com.kurio.mvvmtesting.model.Note

class NoteRepo(application: Application) {
    private val noteDao: NoteDao = NoteRoomDatabase.getInstance(application).noteDao()
    val allNotes: LiveData<List<Note>>

    init {
        allNotes = noteDao.allNote
    }

    fun insertNote(note: Note) {
        InsertAsyncTask(noteDao).execute(note)
    }

    fun updateNote(note: Note) {
        UpdateAsyncTask(noteDao).execute(note)
    }

    fun deleteNote(note: Note) {
        DeleteAsyncTask(noteDao).execute(note)
    }

    private class InsertAsyncTask internal constructor(private val noteDao: NoteDao) : AsyncTask<Note, Void, Void>() {

        override fun onPreExecute() {
            super.onPreExecute()
        }

        override fun doInBackground(vararg notes: Note): Void? {
            noteDao.insert(notes[0])
            return null
        }

        override fun onPostExecute(aVoid: Void?) {
            super.onPostExecute(aVoid)
        }
    }

    private class UpdateAsyncTask internal constructor(private val noteDao: NoteDao) : AsyncTask<Note, Void, Void>() {

        override fun onPreExecute() {
            super.onPreExecute()
        }

        override fun doInBackground(vararg notes: Note): Void? {
            noteDao.update(notes[0])
            return null
        }

        override fun onPostExecute(aVoid: Void?) {
            super.onPostExecute(aVoid)
        }
    }

    private class DeleteAsyncTask internal constructor(private val noteDao: NoteDao) : AsyncTask<Note, Void, Void>() {

        override fun onPreExecute() {
            super.onPreExecute()
        }

        override fun doInBackground(vararg notes: Note): Void? {
            noteDao.delete(notes[0])
            return null
        }

        override fun onPostExecute(aVoid: Void?) {
            super.onPostExecute(aVoid)
        }
    }

}
